### ISSUE TITLE
So many dependencies?

### ISSUE THREAD

**Issue by zentrope** [juxt/bidi#68]
Was thinking about using bidi as I need some routing for both the server and client sides of an app. When I looked at the dependency tree (`lein deps tree`):

    $ lein with-profile production deps :tree

I found this:

```clojure
 [bidi "1.19.0" :exclusions [[commons-codec]]]
   [com.cemerick/url "0.1.1"]
     [pathetic "0.5.0"]
       [com.cemerick/clojurescript.test "0.0.4"]
   [org.clojure/clojurescript "0.0-2371"]
     [com.google.javascript/closure-compiler "v20140625"]
       [args4j "2.0.26"]
       [com.google.code.findbugs/jsr305 "1.3.9"]
       [com.google.javascript/closure-compiler-externs "v20140625"]
       [com.google.protobuf/protobuf-java "2.5.0"]
       [org.json/json "20090211"]
     [org.clojure/data.json "0.2.3"]
     [org.clojure/google-closure-library "0.0-20140718-946a7d39"]
       [org.clojure/google-closure-library-third-party "0.0-20140718-946a7d39"]
     [org.mozilla/rhino "1.7R4"]
   [ring/ring-core "1.2.1"]
     [clj-time "0.4.4"]
       [joda-time "2.1"]
     [commons-fileupload "1.3"]
     [commons-io "2.4"]
     [ring/ring-codec "1.0.0"]
```

My main concern is making an uberjar, which, as a result of bidi, will contain all these things I'm not using (when running from an uberjar), including ClojureScript, clojurescript-test, ring, and so on. Should those be marked as dev dependencies or "provided" or something like that? I suppose I can just go with my own `:exclusions` clauses, but it seems like the library itself shouldn't require those, right?

I can see you've got a `bidi.ring` namespace (is there a clojure-doc generated somewhere?) which I suppose requires ring, but what happened to the "one thing well" story! ;) (I can't tell what `make-handler` does. Return a function that takes a request and .... The routes data has to have function references for handlers, not keywords as shown in the readme example, right?

Anyway.... ;)

**Comment by malcolmsparks** [comment]
I'm completely on board with the spirit of this issue. It's easier said than done, but I've already started considering how to strip down bidi to a smaller core. I plan to do this as part of a larger overhaul including a port to Clojure 1.7's reader macros. Thanks for the prompt.

**Comment by malcolmsparks** [comment]
I've removed the clojure and clojurescript dependencies in the previous release. ring-core will probably be gone in the next release too

**Comment by Skinney** [comment]
Clojurescript is still included when running with the latest bidi. Couldn't this be solved by placing cljsbuild and clojurescript.test plugins in the dev profile?

**Comment by Skinney** [comment]
Just verified that simply moving the top-level plugins into the dev-profile removes bidi's dependency on clojurescript. Unfortunetly, cemerick/url depends on pathetic, which also has cljsbuild as a top level dependency. I've filed an issue with pathetic, and hope that problem will get fixed soon.

**Comment by malcolmsparks** [comment]
Do feel free to send a PR for the clojurescript removal. However, note that
you can use Leiningen's exclusions mechanism to avoid bringing in unwanted
dependencies. Most non-trivial projects need to do that anyway.



Malcolm Sparks
Director

Email: malcolm@juxt.pro
Web: https://juxt.pro

JUXT LTD.
Software Consulting, Delivery, Training

On 9 October 2015 at 05:46, Robin Heggelund Hansen <notifications@github.com

**Comment by Skinney** [comment]
I know, but that doesn't mean libraries shouldn't strive towards a clear seperation between build dependencies, and runtime dependencies. I'll send a PR later today.

**Comment by malcolmsparks** [comment]
I've released 1.23.0 that removes the dependency on cemerick.url - it was only using 2 functions and these were easy to re-write in .cljc
